presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -0,0 +1,382 @@
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_SESSION_PROPERTY;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class SessionPropertyManager
+{
+    private static final JsonCodecFactory JSON_CODEC_FACTORY = new JsonCodecFactory();
+    private final ConcurrentMap<String, SessionProperty<?>> allSessionProperties = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, List<SessionPropertyMetadata<?>>> catalogSessionProperties = new ConcurrentHashMap<>();
+
+    public SessionPropertyManager()
+    {
+        this(new SystemSessionProperties());
+    }
+
+    @Inject
+    public SessionPropertyManager(SystemSessionProperties systemSessionProperties)
+    {
+        this(systemSessionProperties.getSessionProperties());
+    }
+
+    public SessionPropertyManager(List<SessionPropertyMetadata<?>> systemSessionPropertiesSessionProperties)
+    {
+        addSystemSessionProperties(systemSessionPropertiesSessionProperties);
+    }
+
+    public void addSystemSessionProperties(List<SessionPropertyMetadata<?>> systemSessionPropertiesSessionProperties)
+    {
+        systemSessionPropertiesSessionProperties
+                .forEach(this::addSystemSessionProperty);
+    }
+
+    public void addSystemSessionProperty(SessionPropertyMetadata<?> sessionProperty)
+    {
+        requireNonNull(sessionProperty, "sessionProperty is null");
+        addSessionPropertyInternal(Optional.empty(), sessionProperty);
+    }
+
+    public void addConnectorSessionProperties(String catalog, List<SessionPropertyMetadata<?>> sessionProperties)
+    {
+        requireNonNull(catalog, "catalog is null");
+        checkArgument(!catalog.isEmpty() && catalog.trim().equals(catalog), "Invalid catalog name '%s'", catalog);
+        requireNonNull(sessionProperties, "sessionProperties is null");
+
+        checkState(catalogSessionProperties.putIfAbsent(catalog, sessionProperties) == null, "SessionProperties for catalog '%s' are already registered", catalog);
+        for (SessionPropertyMetadata<?> sessionProperty : sessionProperties) {
+            addSessionPropertyInternal(Optional.of(catalog), sessionProperty);
+        }
+    }
+
+    private <T> void addSessionPropertyInternal(Optional<String> catalogName, SessionPropertyMetadata<T> sessionProperty)
+    {
+        SessionProperty<T> value = new SessionProperty<>(catalogName, sessionProperty);
+        allSessionProperties.put(value.getFullyQualifiedName(), value);
+    }
+
+    public SessionPropertyMetadata<?> getSessionPropertyMetadata(String name)
+    {
+        requireNonNull(name, "name is null");
```

Comment:
I think the name needs to be lower cased here. Add a test for uppercase session names used with `SET SESSION` and `RESET SESSION`